### PR TITLE
v1alpha2 API

### PR DESCRIPTION
**Pull request by justinsb**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1183)

### CODE REVIEW

**Comment by justinsb** [comment]
@kris-nova the TF changes are now manageable I think:

* The private subnet cannot have the private- prefix (sadly), because our conversion
logic doesn't have knowledge of the cluster network topology
* ELB has health check
* Bastions now have IAM profile & role
* Renamed bastion_to_master -> all_bastion_to_master
* Renamed SecurityGroupRules that have a CIDR include the CIDR in the name

So if you have a look at the TF changes here I think they are more understandable now: https://github.com/kubernetes/kops/pull/1183/commits/dfc2fa71bc6418d1e08ad2ee80d0729986c033b7

File: tests/integration/privateweave/kubernetes.tf [hunk 1]
```diff
@@ -13,7 +13,7 @@ resource "aws_autoscaling_group" "bastion-privateweave-example-com" {
   launch_configuration = "${aws_launch_configuration.bastion-privateweave-example-com.id}"
   max_size = 1
   min_size = 1
-  vpc_zone_identifier = ["${aws_subnet.private-us-test-1a-privateweave-example-com.id}"]
+  vpc_zone_identifier = ["${aws_subnet.utility-us-test-1a-privateweave-example-com.id}"]
```
Comment (kris-nova) [inline tests/integration/privateweave/kubernetes.tf]: @justinsb 

We aren't renaming `private` subnets to `utility` subnets are we?

In general the 2 conventions that we were using

# Public
These are where the NAT gateways live, and have the synonym **utility**

# Private
These are private, and hold the resources like instances

**Comment by kris-nova** [comment]
@justinsb your changes LGTM if you could please follow up on my comment about the `private` keyword.. just wondering about  the context you are using it - otherwise the TF changes look great

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

**Comment by chrislovecnm** [comment]
Also CI and e2e. Any way we going to get those to pass?

File: Makefile [hunk 2]
```diff
@@ -244,3 +244,15 @@ channels-gocode:
 
 examples:
 	go install k8s.io/kops/examples/kops-api-example/...
+
+# -----------------------------------------------------
+# api machinery regenerate
+
+apimachinery:
+	#go install ./cmd/libs/go2idl/conversion-gen
+	~/k8s/bin/conversion-gen  --input-dirs k8s.io/kops/pkg/apis/kops/v1alpha1 --v=8  --output-file-base=zz_generated.conversion
```
Comment (chrislovecnm) [inline Makefile]: Should we have the comments parts in another target and have this documented?

File: cmd/kops/create_ig.go [hunk 3]
```diff
@@ -96,7 +96,12 @@ func RunCreateInstanceGroup(f *util.Factory, cmd *cobra.Command, args []string,
 	// Populate some defaults
 	ig := &api.InstanceGroup{}
 	ig.ObjectMeta.Name = groupName
-	ig.Spec.Role = api.InstanceGroupRole(options.Role)
+
+	role, ok := api.ParseInstanceGroupRole(options.Role, true)
```
Comment (chrislovecnm) [inline cmd/kops/create_ig.go]: So I got coached on using err, so we should use err ;)

File: cmd/kops/rollingupdate_cluster.go [hunk 4]
```diff
@@ -57,8 +59,10 @@ func init() {
 	cmd.Flags().BoolVar(&rollingupdateCluster.Yes, "yes", false, "perform rolling update without confirmation")
 	cmd.Flags().BoolVar(&rollingupdateCluster.Force, "force", false, "Force rolling update, even if no changes")
 	cmd.Flags().BoolVar(&rollingupdateCluster.CloudOnly, "cloudonly", false, "Perform rolling update without confirming progress with k8s")
+
 	cmd.Flags().DurationVar(&rollingupdateCluster.MasterInterval, "master-interval", 5*time.Minute, "Time to wait between restarting masters")
 	cmd.Flags().DurationVar(&rollingupdateCluster.NodeInterval, "node-interval", 2*time.Minute, "Time to wait between restarting nodes")
+	cmd.Flags().DurationVar(&rollingupdateCluster.BastionInterval, "bastion-interval", 5*time.Minute, "Time to wait between restarting bastions")
```
Comment (chrislovecnm) [inline cmd/kops/rollingupdate_cluster.go]: Oh crap, this will impact my pr ...

File: pkg/apis/kops/cluster.go [hunk 5]
```diff
@@ -273,28 +265,36 @@ type EtcdClusterSpec struct {
...
-	// 1 - Utility (Public) Subnets
-	// 2 - Operating (Private) Subnets
+const (
+	SubnetTypePublic  SubnetType = "Public"
+	SubnetTypePrivate SubnetType = "Private"
+	SubnetTypeUtility SubnetType = "Utility"
+)
+
+type ClusterSubnetSpec struct {
+	// TODO: Rename
```
Comment (chrislovecnm) [inline pkg/apis/kops/cluster.go]: Is this todo still valid?

File: pkg/apis/kops/cluster.go [hunk 6]
```diff
@@ -307,41 +307,17 @@ type ClusterZoneSpec struct {
...
-}
-
 // FillDefaults populates default values.
 // This is different from PerformAssignments, because these values are changeable, and thus we don't need to
 // store them (i.e. we don't need to 'lock them')
 func (c *Cluster) FillDefaults() error {
-	if len(c.Spec.AdminAccess) == 0 {
-		c.Spec.AdminAccess = append(c.Spec.AdminAccess, "0.0.0.0/0")
+	// TODO: Move elsewhere
+	if len(c.Spec.SSHAccess) == 0 {
```
Comment (chrislovecnm) [inline pkg/apis/kops/cluster.go]: Think @vendrov can refactor

File: pkg/apis/kops/parse.go [hunk 7]
```diff
@@ -22,6 +22,27 @@ import (
...
+		s = strings.ToLower(s)
+		if lenient {
+			s = strings.TrimSuffix(s, "s")
+		}
+		if s == findRole {
+			return role, true
+		}
+	}
+	return "", false
+}
```
Comment (chrislovecnm) [inline pkg/apis/kops/parse.go]: Would it make sense to have a point so we can return nil?

File: pkg/apis/kops/v1alpha1/conversion.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
+// +build !ignore_autogenerated
+
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/conversion.go]: And this is API machinery ... :facepalm

File: pkg/apis/kops/v1alpha1/zz_generated.conversion.go [hunk 9]
```diff
@@ -0,0 +1,1182 @@
+// +build !ignore_autogenerated
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/zz_generated.conversion.go]: Is this passing our header check?

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+	//  * enable debugging handlers on the master, so kubectl logs works
+	IsolateMasters *bool `json:"isolateMasters,omitempty"`
+
+	// UpdatePolicy determines the policy for applying upgrades automatically.
+	// Valid values:
+	//   'external' do not apply updates automatically - they are applied manually or by an external system
+	//   missing: default policy (currently OS security upgrades that do not require a reboot)
+	UpdatePolicy *string `json:"updatePolicy,omitempty"`
+
+	//HairpinMode                   string `json:",omitempty"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha2/cluster.go]: This is a todo, but I would like this in comments not in code

File: pkg/apis/kops/v1alpha2/doc.go [hunk 11]
```diff
@@ -0,0 +1,17 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha2/doc.go]: Do we need to start gening go docs for this?

Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha2/doc.go]: I think we get a swagger spec for free

File: pkg/apis/kops/validation.go [hunk 12]
```diff
@@ -319,15 +329,11 @@ func (c *Cluster) Validate(strict bool) error {
...
-	if c.Spec.Topology.Bastion != nil && c.Spec.Topology.Bastion.Enable {
+	if c.Spec.Topology.Bastion != nil {
 		if c.Spec.Topology.Masters == TopologyPublic || c.Spec.Topology.Nodes == TopologyPublic {
 			return fmt.Errorf("Bastion supports only Private Masters and Nodes")
 		}
-		if c.Spec.Topology.Bastion.MachineType == "" {
-			return fmt.Errorf("Bastion MachineType can not be empty")
-		}
-		if c.Spec.Topology.Bastion.IdleTimeout <= 0 {
+		if fi.Int64Value(c.Spec.Topology.Bastion.IdleTimeout) <= int64(0) {
```
Comment (chrislovecnm) [inline pkg/apis/kops/validation.go]: There is an upper end of 300 on this as well.

File: pkg/model/autoscalinggroup.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	NodeUpSource     string
+	NodeUpSourceHash string
+
+	NodeUpConfigBuilder func(ig *kops.InstanceGroup) (*nodeup.NodeUpConfig, error)
+}
+
+var _ fi.ModelBuilder = &AutoscalingGroupModelBuilder{}
+
+func (b *AutoscalingGroupModelBuilder) Build(c *fi.ModelBuilderContext) error {
+	for _, ig := range b.InstanceGroups {
```
Comment (chrislovecnm) [inline pkg/model/autoscalinggroup.go]: Should we be fmt the err's?

File: pkg/model/bastion.go [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+type BastionModelBuilder struct {
+	*KopsModelContext
+}
+
+var _ fi.ModelBuilder = &BastionModelBuilder{}
+
+func (b *BastionModelBuilder) Build(c *fi.ModelBuilderContext) error {
+	if !b.Cluster.IsTopologyPrivate() {
+		return nil
+	}
```
Comment (chrislovecnm) [inline pkg/model/bastion.go]: Again what is your thought on fmt the errors?

File: pkg/model/bastion.go [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+			Zone:               b.LinkToDNSZone(),
+			ResourceType:       s("A"),
+			TargetLoadBalancer: elb,
+		}
+		c.AddTask(t)
+	}
+	return nil
+}
+
+//func (b*BastionModelBuilder) buildASG(ig *kops.InstanceGroup) (*awstasks.AutoscalingGroup, error) {
```
Comment (chrislovecnm) [inline pkg/model/bastion.go]: Can we kill the dead code?

File: pkg/model/names.go [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &awstasks.SSHKey{Name: &sshKeyName}, nil
+}
+
+func (b *KopsModelContext) LinkToSubnet(z *kops.ClusterSubnetSpec) *awstasks.Subnet {
+	//var name string
+	//
```
Comment (chrislovecnm) [inline pkg/model/names.go]: It is dead Jim, please remove ;)

File: pkg/model/sshkey.go [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+	*KopsModelContext
+}
+
+var _ fi.ModelBuilder = &SSHKeyModelBuilder{}
+
+func (b *SSHKeyModelBuilder) Build(c *fi.ModelBuilderContext) error {
+	name, err := b.SSHKeyName()
+	if err != nil {
+		return err
+	}
```
Comment (chrislovecnm) [inline pkg/model/sshkey.go]: Fmt pretty please

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 18]
```diff
@@ -373,19 +385,19 @@ func (c *ApplyClusterCmd) Run() error {
...
 		if role == "" {
-			return "", fmt.Errorf("cannot determine role for instance group: %v", ig.ObjectMeta.Name)
+			return nil, fmt.Errorf("cannot determine role for instance group: %v", ig.ObjectMeta.Name)
 		}
 
 		nodeUpTags, err := buildNodeupTags(role, tf.cluster, tf.tags)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/apply_cluster.go]: SGTM - might I have a fmt please sir

File: upup/pkg/fi/cloudup/awstasks/vpcdhcpoptionsassociation_fitask.go [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+	var jsonName string
+	if err := json.Unmarshal(data, &jsonName); err == nil {
+		o.Name = &jsonName
+		return nil
+	}
+
+	var r realVPCDHCPOptionsAssociation
+	if err := json.Unmarshal(data, &r); err != nil {
+		return err
+	}
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/awstasks/vpcdhcpoptionsassociation_fitask.go]: I think I should stop asking

File: upup/pkg/fi/cloudup/subnets.go [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+	return len(a)
+}
+func (a ByZone) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+func (a ByZone) Less(i, j int) bool {
+	return a[i].Zone < a[j].Zone
+}
+
+func assignCIDRsToSubnets(c *kops.Cluster) error {
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/subnets.go]: This is fun code. How much test coverage do we have?

File: upup/pkg/fi/cloudup/subnets_test.go [hunk 21]
```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/subnets_test.go]: Would love to see some negative test cases. If I am reading this correctly we do not have any, could be wrong

File: upup/pkg/kutil/rollingupdate_cluster.go [hunk 22]
```diff
@@ -105,17 +108,44 @@ func (c *RollingUpdateCluster) RollingUpdate(groups map[string]*CloudInstanceGro
...
 		case api.InstanceGroupRoleMaster:
 			masterGroups[k] = group
+		case api.InstanceGroupRoleBastion:
+			bastionGroups[k] = group
 		default:
 			return fmt.Errorf("unknown group type for group %q", group.InstanceGroup.ObjectMeta.Name)
 		}
 	}
 
+	// Upgrade bastions first; if these go down we can't see anything
```
Comment (chrislovecnm) [inline upup/pkg/kutil/rollingupdate_cluster.go]: Ok so we update the bastion here.  No drain, but definitely validate.

Comment (chrislovecnm) [inline upup/pkg/kutil/rollingupdate_cluster.go]: Actually no validate

File: upup/pkg/fi/cloudup/subnets.go (hunk 20 above)
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/subnets.go]: I see unit tests.. See comment about negative test cases.

File: Makefile (hunk 2 above)
Comment (justinsb) [inline Makefile]: Well... so the challenge is that the fix I had to make it k8s so the conversion generator works only just landed today.  So we also have to update k8s to the latest, or pull it out into the shared repo.  I'm inclined to do the latter.  It's a placeholder for actually installing the code for real

Comment (chrislovecnm) [inline Makefile]: I would encourage us to bump k8s version before this merges.

File: pkg/apis/kops/cluster.go (hunk 5 above)
Comment (justinsb) [inline pkg/apis/kops/cluster.go]: Yes, I probably should do this - I changed it from Name -> SubnetName to force compiler errors, because previous Name was actually Zone.  This rename will not be visible outside the code though.  But I'll do it!

File: pkg/apis/kops/parse.go (hunk 7 above)
Comment (justinsb) [inline pkg/apis/kops/parse.go]: This is a different pattern I think.  It could return (*InstanceGroupRole), but it will never (currently) return an error.  But then we probably want to return (*InstanceGroupRole, error) in case it can return an error in future.  And then we start to realize why exceptions are a good idea ;-)

File: pkg/apis/kops/v1alpha1/conversion.go (hunk 8 above)
Comment (justinsb) [inline pkg/apis/kops/v1alpha1/conversion.go]: Actually even better... this file is _not_ autogenerated, and we already name the autogen files `zz_generated`.  Will fix this one :-)

File: pkg/apis/kops/v1alpha2/cluster.go (hunk 10 above)
Comment (justinsb) [inline pkg/apis/kops/v1alpha2/cluster.go]: We can probably just remove most of these I think ... but a separate PR I think

File: pkg/apis/kops/validation.go (hunk 12 above)
Comment (justinsb) [inline pkg/apis/kops/validation.go]: Cool - want to open an issue / send a PR?

File: pkg/model/autoscalinggroup.go (hunk 13 above)
Comment (justinsb) [inline pkg/model/autoscalinggroup.go]: Not if they are already formatted.  In general if we're calling into our own code, we try to return a formatted error.  But then if we have nothing to add, we don't.

File: pkg/model/bastion.go (hunk 15 above)
Comment (justinsb) [inline pkg/model/bastion.go]: Good call

File: pkg/model/names.go (hunk 16 above)
Comment (justinsb) [inline pkg/model/names.go]: 👍

File: tests/integration/privateweave/kubernetes.tf (hunk 1 above)
Comment (justinsb) [inline tests/integration/privateweave/kubernetes.tf]: We are not, but "normal" public & private subnets (where nodes live) are unqualified i.e. `us-east-1a.k8s.example.com` in both topologies.  We have to do this because when we're doing the version conversion we don't have access to the full cluster state.  It does have some advantages though - I think it might be possible to move from topology public <-> private, though I haven't tried it..

Comment (justinsb) [inline tests/integration/privateweave/kubernetes.tf]: The utility subnet (ELBs, NAT gateways) is only used in private topologies, and is always public.  I did debate renaming it to `dmz` (as I _think_ that's the common term) but utility is probably at least as clear and less likely to offend.

**Comment by chrislovecnm** [comment]
Think this PR will only address a couple of open bugs... SGTM

File: upup/pkg/fi/cloudup/awstasks/vpcdhcpoptionsassociation_fitask.go (hunk 19 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/vpcdhcpoptionsassociation_fitask.go]: Particularly on autogenerated code :-)

File: upup/pkg/fi/cloudup/subnets_test.go (hunk 21 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/subnets_test.go]: Let's do it in another PR

File: pkg/apis/kops/validation.go (hunk 12 above)
Comment (chrislovecnm) [inline pkg/apis/kops/validation.go]: https://github.com/kubernetes/kops/issues/1198

File: upup/pkg/fi/cloudup/awstasks/vpcdhcpoptionsassociation_fitask.go (hunk 19 above)
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/awstasks/vpcdhcpoptionsassociation_fitask.go]: Got me on that one

File: upup/pkg/fi/cloudup/subnets_test.go (hunk 21 above)
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/subnets_test.go]: https://github.com/kubernetes/kops/issues/1199

File: pkg/model/misc.go [hunk 23]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (kris-nova) [inline pkg/model/misc.go]: Convenience functions are good.. and I see when they are useful.. 

But I have to gripes here..

### misc.go

Just naming a file that, opens up the door to contributors dropping in arbitrary logic they don't want to figure out where it should really go

Suggest: `convenience.go` 

### s() and i64()

While these are super handy, and I love them in the new models.. they don't really tell the dev (usually me) what the heck they do..

(Yes I mean this) Can we maybe use the convenience functions as wrappers to very specific verbose function names?

```
func s(v string) *string {
    return stringToStringPointer(v)
}

func stringToStringPointer(v string) *string {
    return &v
}
```

It might seem like super over kill to dig 3 functions deep just to add a `&` to the beginning of a string variable name, but pointers are a **huge** deal. Also when we are working on the models, if we need to see what `s()` does, the code will document itself immediately.

File: tests/integration/create_cluster/ha/expected-v1alpha2.yaml [hunk 24]
```diff
@@ -0,0 +1,118 @@
+apiVersion: kops/v1alpha2
```
Comment (kris-nova) [inline tests/integration/create_cluster/ha/expected-v1alpha2.yaml]: So this is great - I love seeing these new tests come to life - Good job @justinsb

File: upup/pkg/fi/cloudup/subnets.go [hunk 25]
```diff
@@ -0,0 +1,182 @@
+package cloudup
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/subnets.go]: Headers here,

Also GJ on getting the subnets here - you read my mind

File: upup/pkg/kutil/rollingupdate_cluster.go (hunk 22 above)
Comment (kris-nova) [inline upup/pkg/kutil/rollingupdate_cluster.go]: Yep - just get these out of the way first IMHO

File: cmd/kops/rollingupdate_cluster.go (hunk 4 above)
Comment (chrislovecnm) [inline cmd/kops/rollingupdate_cluster.go]: This needed

File: pkg/apis/kops/cluster.go (hunk 6 above)
Comment (kris-nova) [inline pkg/apis/kops/cluster.go]: What needs to be changed here?

File: upup/pkg/kutil/rollingupdate_cluster.go (hunk 22 above)
Comment (chrislovecnm) [inline upup/pkg/kutil/rollingupdate_cluster.go]: Great test for this PR, push my PR to after this monster

File: pkg/model/misc.go (hunk 23 above)
Comment (kris-nova) [inline pkg/model/misc.go]: Just in general whenever a shortcut name like `s()` is used, I have found it useful to have it wrap a more verbose function - that's all I am getting at - simple code doc would work too

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+	VolumeType      *string `json:"volumeType,omitempty"`
+	VolumeSize      *int    `json:"volumeSize,omitempty"`
+	KmsKeyId        *string `json:"kmsKeyId,omitempty"`
+	EncryptedVolume *bool   `json:"encryptedVolume,omitempty"`
+}
+
+// SubnetType string describes subnet types (public, private, utility)
+type SubnetType string
+
+const (
```
Comment (kris-nova) [inline pkg/apis/kops/v1alpha2/cluster.go]: I thought we were using `private` and `utility` without a concept of main..?

File: cmd/kops/rollingupdate_cluster.go (hunk 4 above)
Comment (kris-nova) [inline cmd/kops/rollingupdate_cluster.go]: What is needed? We have nothing here that suggests a change to be honest. 

Don't understand what we need to do

File: pkg/apis/kops/v1alpha1/conversion.go [hunk 27]
```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (kris-nova) [inline pkg/apis/kops/v1alpha1/conversion.go]: > will get its own conversions.go

That is all I needed to read - and we should research how k8s is pulling off the alpha shenanigans - might be useful..

File: pkg/apis/kops/v1alpha1/register.go [hunk 28]
```diff
@@ -60,3 +70,28 @@ func (obj *InstanceGroup) GetObjectKind() unversioned.ObjectKind {
 func (obj *Federation) GetObjectKind() unversioned.ObjectKind {
 	return &obj.TypeMeta
 }
+
+func addConversionFuncs(scheme *runtime.Scheme) error {
+	// Add non-generated conversion functions
+	err := scheme.AddConversionFuncs(
```
Comment (kris-nova) [inline pkg/apis/kops/v1alpha1/register.go]: Touche!

File: cmd/kops/rollingupdate_cluster.go (hunk 4 above)
Comment (chrislovecnm) [inline cmd/kops/rollingupdate_cluster.go]: I will handle on my PR ;) We cannot drain a Bastion ;)

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+	// It cannot overlap ServiceClusterIPRange
+	NonMasqueradeCIDR string `json:"nonMasqueradeCIDR,omitempty"`
+
+	// SSHAccess determines the permitted access to SSH
+	// Currently only a single CIDR is supported (though a richer grammar could be added in future)
+	SSHAccess []string `json:"sshAccess,omitempty"`
+
+	// APIAccess determines the permitted access to the API endpoints (master HTTPS)
+	// Currently only a single CIDR is supported (though a richer grammar could be added in future)
+	APIAccess []string `json:"apiAccess,omitempty"`
```
Comment (kris-nova) [inline pkg/apis/kops/v1alpha2/cluster.go]: Agree

File: pkg/model/external_access.go [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+	if len(b.Cluster.Spec.APIAccess) == 0 {
+		glog.Warningf("APIAccess is empty")
+	}
+
+	if len(b.Cluster.Spec.SSHAccess) == 0 {
+		glog.Warningf("SSHAccess is empty")
+	}
+
+	// SSH is open to AdminCIDR set
+	if b.Cluster.IsTopologyPublic() {
```
Comment (kris-nova) [inline pkg/model/external_access.go]: yes yes indeed - this is far from a deal breaker, and more exciting than anything.. The PRs coming through on this code are going to do just fine - no refactor needed :)

**Comment by justinsb** [comment]
Removing WIP!

**Comment by kris-nova** [comment]
This is a big PR, with a ton of work thanks to @justinsb - This change LGTM

image
